HSL DateTime: Time (interval) class

jjergus opened the pull request:
This is a small stand-alone initial part of the DateTime library (the `HH\Lib\Time` class, which doesn't depend on anything else, and an associated test).

I'd like to get this reviewed before I spend too much time working on the remaining parts. Here I am mainly looking for feedback on:

- Any high-level/general feedback about the proposed API/implementation
  - by "general" I mean anything that I would need to take into account when working on the remaining parts; feedback about specific details like "the toString method is terrible" is welcome as well, but not as important -- specific details can always be changed later, but if anything looks like it's going in some bad general direction, I'd like to know before I go any further in that bad direction
- Is the amount of documentation and test coverage here too little/just about right/too much for an hsl-experimental library?
- Should I separate the rest of the library into similar smaller parts like this, to be reviewed and committed separately, or is it better if I prepare everything and have it reviewed all at once?
  - You can always look at https://github.com/jjergus/hsl-experimental/tree/dt4/src/datetime to see how anything works in context (there's also README.md there explaining most of the design decisions).
- Do we need to decouple hsl-experimental from FB WWW before I start committing these? I think we probably don't want to have these in the `HH\Lib` namespace in FB WWW, but either of these seems reasonable to me:
  - put these in `HH\Lib\Experimental` namespace and commit to WWW
  - keep these in `HH\Lib` but decouple hsl-experimental from WWW first
  - or even both: decouple hsl-experimental from WWW and also move these to `HH\Lib\Experimental`
  - having these in FB WWW would mean people could start using them, which has disadvantages (hard to change things later), but also advantages (much better chance to get useful contributions from FB engineers? I expect the first person/team that wants to use it would effectively become the main maintainer, which could be better than it effectively becoming unmaintained when I stop working on it)

azjezz commented on src/datetime/Time.php:
```diff
@@ -0,0 +1,481 @@
+    } else if ($this->isZero()) {
+      return $other;
+    }
+    return self::hours(
```
name seems misleading if it can take minutes, seconds, ad nanoseconds, should be "create" or "fromPrimitives"

jjergus replied:
Good feedback! I wasn't 100% sure about this naming pattern, but my general idea was to optimize for the common case (which is a single argument  like `Time::hours(42)`), then having the option to provide the additional arguments is just a bonus on top of that (users can also choose to avoid it if they find it confusing I guess).

In my native language it's also reasonable to write something like "3:30 hours" which is understood to mean "3 hours and 30 minutes", but I'm not sure if this applies to English as well :)

`create`/`fromPrimitives`/`fromParts` (or just having a public constructor) could be a good replacement for the current version of `hours` (maybe we can support both? though I have mixed feelings about having too many ways to do the same thing) -- but I think what's nice about the current version is that there is a nice way to create an instance with any combinations of parts.

azjezz replied:
personally, i would prefer: `Time\Interval::empty()->withDays(3)->withHours(12)->withMinutes(32)->withSeconds(31)`

also, Tokio time is a good reference here: https://docs.rs/tokio/1.6.1/tokio/time/index.html

jjergus replied:
That's...very verbose, I don't think I would want to write that regularly, but it works with the current version so you have that option too if you prefer it (performance might be a bit of a concern though, since if we want to keep the objects immutable, that would have to create multiple throwaway intermediate objects).

fredemmott replied:
![image](https://user-images.githubusercontent.com/360927/124008390-fc40cf80-d9a1-11eb-8e04-329d979edbb7.png)

Though perhaps the optional args should go from all of the current ones? I'm don't think `::minutes(12, 30, 100)` is clear when reading either

fredemmott commented:
I'll look more thoroughly through soon, but before I forget:

> keep these in HH\Lib but decouple hsl-experimental from WWW first

Raw `HH\Lib\` should really be preserved for built-ins only. It is **a lot** of work and pain to move stuff from external to built-in without changing the namespace, and essentially forces a minimum  3-week delay to the eventual merge process, to accommodate our need to be able to test/run on older versions of HHVM for a while.

fredemmott commented on src/datetime/Time.php:
```diff
@@ -0,0 +1,482 @@
+   * meant to be a comprehensive way to format time intervals for user-facing
+   * output.
+   */
+  public function toString(int $max_decimals = 3): string {
```
Perhaps implement  `__debugInfo()` instead?

fredemmott commented on src/datetime/_Private/util.php:
```diff
@@ -0,0 +1,23 @@
+ *
+ */
+
+namespace HH\Lib\Experimental\DateTime\_Private;
```
We've been moving to `HH\Lib\Experimental\_Private\_DateTime` in newer stuff (I think this was originally @lexidor 's suggestion? :) )

This lets implementations easily and clearly pull in multiple private namespaces, e.g. `use namespace HH\Lib\_Private\{_DateTime,_OS};`, and makes grepping easier (as `_DateTime` can be assumed to be the private datetime NS, but `_Private` might be something else in other files)

lexidor replied:
I have found the original conversation about this. The suggestion was yours, but I was very enthusiastic about your proposal. It is a nice best practice for files which use `_OS` and `OS`. No need to alias a use clause.

lexidor commented:
Note to self, have a look at dt4 shortly. The API has changed since dt2.

jjergus commented:
- moved to `...\_Private\_DateTime`
- added `fromParts`
- removed optional arguments
- added `__debugInfo()` (but I kept `__toString()` as well, it seems more generally useful to me)
